source/titback/interpreter.cpp

```diff
@@ -0,0 +1,172 @@
+
+#include <cstdio>
+#include <format>
+#include <memory>
+#include <utility>
+
+#include <Python.h> // IWYU pragma: keep
+
+#include "tit/core/checks.hpp"
+#include "tit/core/exception.hpp"
+#include "tit/core/main_func.hpp"
+#include "tit/core/str_utils.hpp"
+#include "tit/core/sys/utils.hpp"
+
+#include "titback/interpreter.hpp"
+
+namespace tit::back {
+
+// NOLINTBEGIN(*-include-cleaner)
+
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+PythonConfig::PythonConfig() : config_{new PyConfig{}} {
+  PyConfig_InitIsolatedConfig(config_.get());
+}
+
+void PythonConfig::Cleaner_::operator()(PyConfig* config) const {
+  if (config != nullptr) PyConfig_Clear(config);
+  std::default_delete<PyConfig>::operator()(config);
+}
+
+auto PythonConfig::base() const noexcept -> PyConfig* {
+  TIT_ASSERT(config_ != nullptr, "Config is not initialized!");
+  return config_.get();
+}
+
+void PythonConfig::set_home(CStrView home) const {
+  const auto status =
+      PyConfig_SetBytesString(base(), &base()->home, home.c_str());
+  if (PyStatus_IsError(status) == 0) return;
+  TIT_THROW("Failed to set Python home directory to '{}': {}: {}.",
+            home,
+            status.func,
+            status.err_msg);
+}
+
+void PythonConfig::set_prog_name(CStrView name) const {
+  const auto status =
+      PyConfig_SetBytesString(base(), &base()->program_name, name.c_str());
+  if (PyStatus_IsError(status) == 0) return;
+  TIT_THROW("Failed to set Python program name to '{}': {}: {}.",
+            name,
+            status.func,
+            status.err_msg);
+}
+
+void PythonConfig::set_cmd_args(CmdArgs args) const {
+  const auto status = PyConfig_SetBytesArgv(
+      base(),
+      static_cast<Py_ssize_t>(args.size()),
+      const_cast<char* const*>(args.data())); // NOLINT(*-const-cast)
+  if (PyStatus_IsError(status) == 0) return;
+  TIT_THROW("Failed to set Python command line arguments: {}: {}.",
+            status.func,
+            status.err_msg);
+}
+
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+PythonInterpreter::PythonInterpreter(PythonConfig config)
+    : config_{std::move(config)} {
+  // Initialize the Python interpreter.
+  const auto status = Py_InitializeFromConfig(config_.base());
+  if (PyStatus_IsError(status) != 0) {
+    TIT_THROW("Failed to initialize Python interpreter: {}: {}.",
+              status.func,
+              status.err_msg);
+  }
+
+  // Get the globals of the main module.
+  auto* const main_module = PyImport_AddModule("__main__");
+  if (main_module == nullptr) {
+    PyErr_Print();
+    TIT_THROW("Failed to import the main module.");
+  }
+  globals_ = PyModule_GetDict(main_module);
+  if (globals_ == nullptr) {
+    PyErr_Print();
+    TIT_THROW("Failed to get the main module globals.");
+  }
+
+  // Initialize the coverage report.
+#ifdef TIT_HAVE_GCOV
+  exec(R"PY(if True:
```

Comment:
Let's remove the first line if it contains only whitespace characters instead of writing `if True:` everywhere.